src/ucs/datastruct/piecewise_func.c

```diff
@@ -0,0 +1,186 @@
+            return;
+        }
+    }
+
+    ucs_bug("List doesn't cover [0, SIZE_MAX] range");
+#endif
+}
+
+double ucs_piecewise_func_apply(const ucs_piecewise_func_t *func, size_t x)
+{
+    ucs_piecewise_segment_t *seg;
+
+    ucs_piecewise_func_assert_covered_range(func);
+
+    ucs_list_for_each(seg, &func->head, list) {
+        if (seg->end >= x) {
+            break;
+        }
+    }
+    return ucs_linear_func_apply(seg->func, x);
+}
+
+static ucs_status_t
+ucs_piecewise_segment_split(ucs_piecewise_segment_t *seg, size_t split_point)
+{
+    ucs_piecewise_segment_t *new_seg;
+
+    ucs_assertv(split_point < seg->end,
+                "%zu < %zu (split point shouldn't cross segment border)",
+                split_point, seg->end);
+
+    new_seg = ucs_malloc(sizeof(ucs_piecewise_segment_t), "piecewise segment");
+    if (new_seg == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_list_insert_after(&seg->list, &new_seg->list);
+    new_seg->end  = seg->end;
+    new_seg->func = seg->func;
+    seg->end      = split_point;
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_piecewise_func_add_range(ucs_piecewise_func_t *func,
+                                          size_t start, size_t end,
+                                          ucs_linear_func_t trend)
+{
+    size_t seg_start = 0;
+    ucs_piecewise_segment_t *seg;
+    ucs_status_t status;
+
+    ucs_piecewise_func_assert_covered_range(func);
+    ucs_assert(start <= end);
+
+    /* Iterate until first affected segment */
+    ucs_list_for_each(seg, &func->head, list) {
+        if (start <= seg->end) {
+            break;
+        }
+        seg_start = seg->end + 1;
+    }
+
+    /* Split the first segment */
+    if (seg_start != start) {
+        if (seg->end > end) {
+            /* The added segment is in the middle of the `seg`,
+             * so it would split it to three parts
+             */
+            status = ucs_piecewise_segment_split(seg, end);
+            if (status != UCS_OK) {
+                return status;
+            }
+        }
+
+        status = ucs_piecewise_segment_split(seg, start - 1);
+        if (status != UCS_OK) {
+            return status;
+        }
+        /* Switch to next segment to not update the original segment by the
+         * added one on the next step
+         */
+        seg = ucs_list_next(&seg->list, ucs_piecewise_segment_t, list);
+    }
+
+    /* Update the segments that are fully covered by added segment
+     * (including the segment splitted the step before)
+     */
+    while (end > seg->end) {
+        ucs_linear_func_add_inplace(&seg->func, trend);
+        seg = ucs_list_next(&seg->list, ucs_piecewise_segment_t, list);
+    }
+
+    /* Update the last segment */
+    if (seg->end != end) {
+        /* Split the last segment */
+        ucs_assert(end != SIZE_MAX);
+
+        status = ucs_piecewise_segment_split(seg, end);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
```

Comment:
maybe we can do the whole logic in single loop? i.e iterating on all segments, split if needed, apply linear func if needed.